modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreCacheGroupProcess.java

```diff
@@ -0,0 +1,647 @@
+        }
+
+        /** @return Snapshot name. */
+        protected String snapshotName() {
+            return snpName;
+        }
+
+        /**
+         * @return List of cache group names to restore from the snapshot.
+         */
+        protected Set<String> groups() {
+            return grps.keySet();
+        }
+
+        /**
+         * @param name Cache name.
+         * @return {@code True} if the cache with the specified name is currently being restored.
+         */
+        protected boolean containsCache(String name) {
+            return cacheIds.contains(CU.cacheId(name));
+        }
+
+        /** @return Cache configurations. */
+        protected Collection<StoredCacheData> configs() {
+            return ccfgs;
+        }
+
+        /**
+         * @param err Error.
+         * @return {@code True} if operation has been interrupted by this call.
+         */
+        protected boolean interrupt(Exception err) {
+            return errRef.compareAndSet(null, err);
+        }
+
+        /**
+         * @return Interrupted flag.
+         */
+        protected boolean interrupted() {
+            return error() != null;
+        }
+
+        /**
+         * @return Error if operation was interrupted, otherwise {@code null}.
+         */
+        protected @Nullable Throwable error() {
+            return errRef.get();
+        }
+
+        /**
+         * Restore specified cache groups from the local snapshot directory.
+         *
+         * @param updateMetadata Update binary metadata flag.
+         * @throws IgniteCheckedException If failed.
+         */
+        protected void restore(boolean updateMetadata) throws IgniteCheckedException {
+            if (interrupted())
+                return;
+
+            IgniteSnapshotManager snapshotMgr = ctx.cache().context().snapshotMgr();
+            String folderName = ctx.pdsFolderResolver().resolveFolders().folderName();
+
+            if (updateMetadata) {
+                File binDir = binaryWorkDir(snapshotMgr.snapshotLocalDir(snpName).getAbsolutePath(), folderName);
+
+                if (!binDir.exists()) {
+                    throw new IgniteCheckedException("Unable to update cluster metadata from snapshot, " +
+                        "directory doesn't exists [snapshot=" + snpName + ", dir=" + binDir + ']');
+                }
+
+                ctx.cacheObjects().updateMetadata(binDir, this::interrupted);
+            }
+
+            for (File grpDir : snapshotMgr.snapshotCacheDirectories(snpName, folderName)) {
+                String grpName = FilePageStoreManager.cacheGroupName(grpDir);
+
+                if (!groups().contains(grpName))
+                    continue;
+
+                snapshotMgr.restoreCacheGroupFiles(snpName, grpName, grpDir, this::interrupted, grps.get(grpName));
+            }
+        }
+
+        /**
+         * Rollback changes made by process in specified cache group.
+         */
+        protected void rollback() {
```

Comment:
We also should not place such methods in the context. Let's pass the `ctx` into this method if we should do any actions.